go/libraries/doltcore/diff/diffsplitter.go

```diff
@@ -40,128 +56,172 @@ type RowDiff struct {
+	last := query[len(query)-1]
+	if last.Name != "diff_type" {
+		return nil, errors.New("expected last diff column to be 'diff_type'")
+	}
+	query = query[:len(query)-1]
+
+	mapping = make([]int, len(query))
+	for i, col := range query {
+		switch {
+		case strings.HasPrefix(col.Name, fromPrefix):
+			// map "from_..." column to "to_..." column
+			base := col.Name[len(fromPrefix):]
+			mapping[i] = query.IndexOfColName(toPrefix + base)
+		case strings.HasPrefix(col.Name, toPrefix):
+			// map "to_..." column to "from_..." column
+			base := col.Name[len(toPrefix):]
+			mapping[i] = query.IndexOfColName(fromPrefix + base)
+		default:
+			return nil, errors.New("expected column prefix of 'to_' or 'from_' (" + col.Name + ")")
 		}
 	}
+	// |mapping| will contain -1 for unmapped columns
+	return
+}
+
+func (ds DiffSplitter) SplitDiffResultRow(row sql.Row) (from, to RowDiff, err error) {
+	from = RowDiff{ColDiffs: make([]ChangeType, len(ds.targetSch))}
+	to = RowDiff{ColDiffs: make([]ChangeType, len(ds.targetSch))}
+
+	diffType := row[len(row)-1]
+	row = row[:len(row)-1]
+
+	switch diffType.(string) {
+	case removedStr:
+		from.Row = make(sql.Row, len(ds.targetSch))
+		from.RowDiff = Removed
+		for i := 0; i < ds.splitIdx; i++ {
+			j := ds.queryToTarget[i]
+			from.Row[j] = row[i]
+			from.ColDiffs[j] = Removed
+		}
 
-	if diffTypeStr == "added" || diffTypeStr == "modified" {
-		newRow.Row = make(sql.Row, len(ds.targetSch))
-		if diffTypeStr == "modified" {
-			newRow.RowDiff = ModifiedNew
-		} else {
-			newRow.RowDiff = Added
+	case addedStr:
+		to.Row = make(sql.Row, len(ds.targetSch))
+		to.RowDiff = Added
+		for i := ds.splitIdx; i < len(row); i++ {
+			j := ds.queryToTarget[i]
+			to.Row[j] = row[i]
+			to.ColDiffs[j] = Added
 		}
 
-		for i := ds.fromLen; i < len(ds.diffQuerySch)-1; i++ {
-			cmp := ds.diffQuerySch[i].Type.Compare
-			newRow.Row[ds.queryToTarget[i]] = row[i]
-
-			if diffTypeStr == "modified" {
-				// need this to compare map[string]interface{} and other incomparable result types
-				if n, err := cmp(row[i], row[ds.toFrom[i]]); err != nil {
-					return RowDiff{}, RowDiff{}, err
-				} else if n != 0 {
-					newRow.ColDiffs[ds.queryToTarget[i]] = ModifiedNew
-				}
+	case modifiedStr:
+		from.Row = make(sql.Row, len(ds.targetSch))
+		from.RowDiff = ModifiedOld
+		for i := 0; i < ds.splitIdx; i++ {
+			j := ds.queryToTarget[i]
+			from.Row[j] = row[i]
+		}
+		to.Row = make(sql.Row, len(ds.targetSch))
+		to.RowDiff = ModifiedNew
+		for i := ds.splitIdx; i < len(row); i++ {
+			j := ds.queryToTarget[i]
+			to.Row[j] = row[i]
+		}
+		// now do field-wise comparison
+		var cmp int
+		for i, col := range ds.targetSch {
+			cmp, err = col.Type.Compare(from.Row[i], to.Row[i])
+			if err != nil {
+				return RowDiff{}, RowDiff{}, err
+			} else if cmp != 0 {
+				from.ColDiffs[i] = ModifiedOld
+				to.ColDiffs[i] = ModifiedNew
 			} else {
-				newRow.ColDiffs[ds.queryToTarget[i]] = Added
+				from.ColDiffs[i] = None
```

Comment:
Isn't this the zero value?

Reply:
yes, but I thought it better to be explicit